File: packages/progress/src/trend.ts
import type { SessionEvent, Phase, TrendAnalysis } from './types'
import type { HierarchicalResult } from './bayesian'
import type { RegimeShiftResult } from './cusum'
import { runKalmanOnPhase } from './kalman'
import { MIN_SESSIONS_PER_BUCKET } from './config'

export function analyzeTrend(
  sessions: SessionEvent[],
  exceptionalIds: Set<string>,
  bayesianResult: HierarchicalResult,
  cusumResult: RegimeShiftResult,
): TrendAnalysis {
  const activeSessions = sessions.filter(
    (s) =>
      s.source === 'active' &&
      s.plannedMinutes != null &&
      s.plannedMinutes > 0 &&
      s.activeMinutes != null &&
      s.activeMinutes > 0 &&
      (!s.sessionId || !exceptionalIds.has(s.sessionId)),
  )

  if (activeSessions.length === 0) {
    return {
      phases: [],
      currentPhase: null,
      projectionSlope: 0,
      projectionUncertainty: 1,
    }
  }

  const paceRatios = activeSessions.map(
    (s) => s.activeMinutes! / s.plannedMinutes!,
  )

  const mean = paceRatios.reduce((s, r) => s + r, 0) / paceRatios.length
  const sumSq = paceRatios.reduce((s, r) => s + (r - mean) ** 2, 0)
  const measurementVariance =
    paceRatios.length > 1
      ? Math.max(sumSq / (paceRatios.length - 1), 0.001)
      : 0.01

  const breakpoints = cusumResult.breakpoints
  const segmentStarts = [0, ...breakpoints.map((bp) => bp + 1).filter((s) => s < activeSessions.length)]
  const segmentEnds = [...breakpoints.filter((bp) => bp < activeSessions.length), activeSessions.length - 1]

  const segments: Array<[number, number]> = []
  for (let i = 0; i < segmentStarts.length; i++) {
    const start = segmentStarts[i]
    const end = i < segmentEnds.length ? segmentEnds[i] : activeSessions.length - 1
    if (start <= end) {
      segments.push([start, end])
    }
  }

  if (segments.length === 0) {
    segments.push([0, activeSessions.length - 1])
  }

  const phases: Phase[] = segments.map(([start, end]) => {
    const segmentRatios = paceRatios.slice(start, end + 1)
    const segmentSessions = activeSessions.slice(start, end + 1)

    if (segmentRatios.length < MIN_SESSIONS_PER_BUCKET) {
      const segMean =
        segmentRatios.reduce((s, r) => s + r, 0) / segmentRatios.length
      return {
        startSessionIndex: start,
        endSessionIndex: end,
        startDate: segmentSessions[0].date,
        endDate: segmentSessions[segmentSessions.length - 1].date,
        level: segMean,
        slope: 0,
        slopeUncertainty: 1,
        sessionCount: segmentRatios.length,
      }
    }

    const kalmanResult = runKalmanOnPhase(
      segmentRatios,
      bayesianResult.globalPosterior.mean,
      bayesianResult.globalPosterior.variance,
      measurementVariance,
    )

    return {
      startSessionIndex: start,
      endSessionIndex: end,
      startDate: segmentSessions[0].date,
      endDate: segmentSessions[segmentSessions.length - 1].date,
      level: kalmanResult.finalLevel,
      slope: kalmanResult.finalSlope,
      slopeUncertainty: kalmanResult.slopeUncertainty,
      sessionCount: segmentRatios.length,
    }
  })

  const currentPhase = phases.length > 0 ? phases[phases.length - 1] : null

  return {
    phases,
    currentPhase,
    projectionSlope: currentPhase?.slope ?? 0,
    projectionUncertainty: currentPhase?.slopeUncertainty ?? 1,
  }
}

File: college/mydeliverables/phase1-ESA/for-submit/code/packages/py-progress/src/py_progress/trend.py
from __future__ import annotations

from py_progress.config import MIN_SESSIONS_PER_BUCKET
from py_progress.kalman import run_kalman_on_phase
from py_progress.types import (
    HierarchicalResult,
    Phase,
    RegimeShiftResult,
    SessionEvent,
    TrendAnalysis,
)


def analyze_trend(
    sessions: list[SessionEvent],
    exceptional_ids: set[str],
    bayesian_result: HierarchicalResult,
    cusum_result: RegimeShiftResult,
) -> TrendAnalysis:
    active_sessions = [
        s
        for s in sessions
        if s.get("source") == "active"
        and s.get("plannedMinutes") is not None
        and s["plannedMinutes"] > 0
        and s.get("activeMinutes") is not None
        and s["activeMinutes"] > 0
        and (not s.get("sessionId") or s["sessionId"] not in exceptional_ids)
    ]

    if not active_sessions:
        return TrendAnalysis(
            phases=[],
            currentPhase=None,
            projectionSlope=0.0,
            projectionUncertainty=1.0,
        )

    pace_ratios = [s["activeMinutes"] / s["plannedMinutes"] for s in active_sessions]

    mean = sum(pace_ratios) / len(pace_ratios)
    sum_sq = sum((r - mean) ** 2 for r in pace_ratios)
    measurement_variance = (
        max(sum_sq / (len(pace_ratios) - 1), 0.001)
        if len(pace_ratios) > 1
        else 0.01
    )

    breakpoints = cusum_result.breakpoints
    segment_starts = [0] + [
        bp + 1 for bp in breakpoints if bp + 1 < len(active_sessions)
    ]
    segment_ends = [
        bp for bp in breakpoints if bp < len(active_sessions)
    ] + [len(active_sessions) - 1]

    segments: list[tuple[int, int]] = []
    for i, start in enumerate(segment_starts):
        end = segment_ends[i] if i < len(segment_ends) else len(active_sessions) - 1
        if start <= end:
            segments.append((start, end))

    if not segments:
        segments.append((0, len(active_sessions) - 1))

    phases: list[Phase] = []
    for start, end in segments:
        segment_ratios = pace_ratios[start : end + 1]
        segment_sessions = active_sessions[start : end + 1]

        if len(segment_ratios) < MIN_SESSIONS_PER_BUCKET:
            seg_mean = sum(segment_ratios) / len(segment_ratios)
            phases.append(
                Phase(
                    startSessionIndex=start,
                    endSessionIndex=end,
                    startDate=segment_sessions[0]["date"],
                    endDate=segment_sessions[-1]["date"],
                    level=seg_mean,
                    slope=0.0,
                    slopeUncertainty=1.0,
                    sessionCount=len(segment_ratios),
                )
            )
        else:
            kalman_result = run_kalman_on_phase(
                segment_ratios,
                bayesian_result.globalPosterior.mean,
                bayesian_result.globalPosterior.variance,
                measurement_variance,
            )
            phases.append(
                Phase(
                    startSessionIndex=start,
                    endSessionIndex=end,
                    startDate=segment_sessions[0]["date"],
                    endDate=segment_sessions[-1]["date"],
                    level=kalman_result.finalLevel,
                    slope=kalman_result.finalSlope,
                    slopeUncertainty=kalman_result.slopeUncertainty,
                    sessionCount=len(segment_ratios),
                )
            )

    current_phase = phases[-1] if phases else None

    return TrendAnalysis(
        phases=phases,
        currentPhase=current_phase,
        projectionSlope=current_phase.slope if current_phase else 0.0,
        projectionUncertainty=current_phase.slopeUncertainty if current_phase else 1.0,
    )
